rclcpp/include/rclcpp/strategies/message_pool_memory_strategy.hpp

```diff
@@ -87,24 +99,68 @@ class MessagePoolMemoryStrategy
    */
   void return_message(std::shared_ptr<MessageT> & msg)
   {
-    for (size_t i = 0; i < Size; ++i) {
-      if (pool_[i].msg_ptr_ == msg) {
-        pool_[i].used = false;
-        return;
+    (void)msg;
+
+    // What we really want to do here is to figure out whether the user has taken an additional
+    // reference to the message, and only add it to the free list if that is *not* the case.
+    // However, we can't really do that for the currently passed-in msg; it can have an arbitrary
+    // reference count due to the mechanisms of rclcpp.  Instead, we look at all the rest of the
+    // pointers, and add the ones that the user has released into the free pool.
+    // We do the same thing in borrow_message(), so if the user has a pool of size 1
+    // (or only one free slot), we'll always find it.
```

Comment:
I'm not convinced by this strategy, specifically I don't think we should be calling `return_message` when the given message isn't eligible for return. I'd go so far as to say, if `msg.use_count()` is not equal to `1` we throw here. The implication of this is that we should rely on the destructor being called when use_count reaches zero, which we're not doing here. I assume that's to do with not wanting to propagate the custom destructor? Or is it something else?

Reply:
> I assume that's to do with not wanting to propagate the custom destructor? Or is it something else?

It has to do with the interface between the Executor class and the message strategies.

In particular, in [Executor::execute_subscription](https://github.com/ros2/rclcpp/blob/7f7ffcf6ed1143b4d4a9e0cf248e190731b2e5f8/rclcpp/src/rclcpp/executor.cpp#L615) you have this sequence:

```
          std::shared_ptr<void> message = subscription->create_message();
          take_and_do_error_handling(
            "taking a message from topic",
            subscription->get_topic_name(),
            [&]() {return subscription->take_type_erased(message.get(), message_info);},
            [&]() {subscription->handle_message(message, message_info);});
          subscription->return_message(message);
```

And the thing is, you *have* to do that sequence for the default message strategy of allocation.  That is, you need to "borrow" (allocate), pass that to the user callback, and then "return" (deallocate).  For a user callback with ConstRef type, this does exactly what we expect.  For a user callback with a shared_ptr type, the user may have incremented the reference count, in which case the "return" won't actually deallocate, but will instead decrement the reference count.  But either way, you have to do that decrement, otherwise you'll have a memory leak.

So with that interface in mind, you now come to this message pool, which has to fit into the same interface.  The "borrow" doesn't do an allocation, but instead just pulls the message from a preallocated pool.  The "return" really wants to return the message to the pool, but there are several internal rclcpp references to the shared_ptr at that time; one from the pool itself, one from `Executor::execute_subscription`, and one from https://github.com/ros2/rclcpp/blob/7f7ffcf6ed1143b4d4a9e0cf248e190731b2e5f8/rclcpp/include/rclcpp/subscription.hpp#L371 .

I originally implemented this with a check that returns the current message to the pool if `use_count() == 3`, and that works.  But it seemed really fragile to rely on that implementation detail (though we could mitigate that fragility with some tests).  That's why I ended up with this strategy of not checking the current message, but checking everything else in the pool instead (since by that point, all of the internal rclcpp references should have been dropped).

Does this make sense?

Reply:
how about letting the user application know this situation at least once with warning message? i think it would be better to do that for user and debugging purpose instead of being silent?

Reply:
> how about letting the user application know this situation at least once with warning message?

Can you please clarify which situation you think we should warn about?  Thanks.

Reply:
I was thinking something like this would be better? instead of being silent.

```suggestion
    if (msg.use_count() >= 3) {
      RCLCPP_WARN_ONCE(
        rclcpp::get_logger("rclcpp"),
        "User application still keeps the reference to the message.",
        "The message memory cannot be reused until user frees the reference to the message.");
```

Reply:
> I was thinking something like this would be better? instead of being silent.

Oh, right.  Well, it is a bit tricky.  In some sense, if the user holds onto a shared_ptr, they should know what they are doing and eventually release it.  It is kind of a limitation/feature of C++.  So my vote would not be to warn in this situation.